File: leetcode/2.两数相加.cpp
#include <vector>

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

using namespace std;
class Solution {
public:
    static ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int step = 0;
        auto pNode = new ListNode(0);
        ListNode* node = pNode;
        for (;l1 != nullptr || l2 != nullptr;) {
            int l1v = 0;
            int l2v = 0;
            if (l1) {
                l1v = l1->val;
                l1 = l1->next;
            }

            if (l2)
            {
                l2v = l2->val;
                l2 = l2->next;
            }

            int res = l1v + l2v + step;
            step = res / 10;
            res %= 10;
            node->val = res;

            if (l1 != nullptr || l2 != nullptr || step) {
                node->next = new ListNode(0);
                node = node->next;
            }

        }
        if (step) {
            node->val = step;
        }

        return pNode;
    }
};

int main() {
    ListNode l1 = ListNode(0);
    ListNode l2 = ListNode(0);
    l1.val = 6;
    l2.val = 7;
    ListNode* l3 = Solution::addTwoNumbers(&l1, &l2);
    return 0;
}

File: leetcode/2.两数相加.rs
//----------------------------------------------------------------------------------
type ListNodePtr2 = Box<ListNode2>;

struct ListNode2 {
    val: i32,
    next: Option<ListNodePtr2>,
}

impl ListNode2 {
    fn new(val: i32) -> ListNode2 {
        ListNode2 {
            val,
            next: None,
        }
    }

    fn prepare(mut self, val: i32) -> ListNode2 {
        let mut cur_node = &mut self;
        while cur_node.next.is_some() {
            cur_node  = cur_node.next.as_mut().unwrap().as_mut();
        }
        cur_node.next = Some(Box::new(ListNode2::new(val)));
        self
    }
}

fn two_num_add2(mut l1: Option<ListNodePtr2>, mut l2: Option<ListNodePtr2>) -> Option<ListNodePtr2> {
    let mut step:i32 = 0;
    let mut res = Box::new(ListNode2::new(step));
    let mut cur_node = res.as_mut();

    while l1.is_some() || l2.is_some() {


        let mut l1v = 0;
        let mut l2v = 0;
        if let Some(l1r) = l1 {
            l1v = l1r.val;
            l1 = l1r.next;
        }

        if let Some(l2r) = l2 {
            l2v = l2r.val;
            l2 = l2r.next;
        }

        let mut res_val = l1v + l2v + step;
        step = res_val / 10;
        res_val %= 10;

        cur_node.val = res_val;

        if l1.is_some() || l2.is_some() || step != 0 {
            let next_node = Box::new(ListNode2::new(step));
            cur_node.next = Some(next_node);
            cur_node = cur_node.next.as_mut().unwrap().as_mut();
        }
    };
    Some(res)
}
//----------------------------------------------------------------------------------


fn main() {

    let node_a_list2 = ListNode2::new(1).prepare(2).prepare(3);
    let node_b_list2 = ListNode2::new(4).prepare(5).prepare(9);
    let _res_list2 = two_num_add2(Some(Box::new(node_a_list2)), Some(Box::new(node_b_list2)));
    println!("finish!");
}

